File: Simulation-Results_Files/simulation_storm/storm/debug_plot.R
library(duckdb)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(scales)

# Adjust plot theme globally
theme_set(theme_bw(base_size = 14) + 
            theme(text = element_text(family = "serif", face = "plain")))

# Parse the '3tot' column and extract values based on keys
parse_3tot <- function(df, y_col_series, z_col_series) {
  extract_values <- function(series) {
    y_vals <- numeric(length(series))
    z_vals <- numeric(length(series))
    
    for (i in seq_along(series)) {
      parts <- strsplit(as.character(series[i]), ",")[[1]]
      for (part in parts) {
        matches <- regmatches(part, regexec("([0-9.-]+)\\((-?[0-9]+)\\)", part))
        if (length(matches[[1]]) > 1) {
          value <- as.numeric(matches[[1]][2])
          key <- as.integer(matches[[1]][3])
          if (key == y_col_series[i]) {
            y_vals[i] <- value
          } else if (key == z_col_series[i]) {
            z_vals[i] <- value
          }
        }
      }
    }
    
    list(y = y_vals, z = z_vals)
  }
  
  vals <- extract_values(df$`3tot`)
  df %>% mutate(`3tot_y` = vals$y, `3tot_z` = vals$z)
}

# Fetch and process data
fetch_and_process_data <- function(database, rep, gen) {
  print(sprintf("Fetching data for generation %s from %s", gen, database))
  conn <- dbConnect(duckdb::duckdb(), dbdir = database, read_only = TRUE)
  on.exit(dbDisconnect(conn))
  
  df <- dbGetQuery(conn, sprintf("SELECT y, z, '3tot', gen FROM simulations WHERE popstat = 'ok' AND rep = %d AND gen = %d ORDER BY gen;", rep, gen))
  if (nrow(df) > 0) {
    parsed_data <- parse_3tot(df, df$y, df$z)
    summarise(group_by(parsed_data, y, z), `3tot_y` = mean(`3tot_y`, na.rm = TRUE), `3tot_z` = mean(`3tot_z`, na.rm = TRUE))
  } else {
    data.frame()
  }
}

# Plot data for all generations
plot_data <- function(database, output_file) {
  generations <- seq(1, 5000, by = 500)
  plots <- lapply(generations, function(gen) {
    gen_data <- fetch_and_process_data(database, 1, gen)
    if (nrow(gen_data) > 0) {
      ggplot(gen_data, aes(x = z, y = y, fill = (`3tot_y` - `3tot_z`) / (`3tot_y` + `3tot_z`))) +
        geom_tile() +
        scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0) +
        labs(title = sprintf("Gen %s", gen))
    } else {
      ggplot() + labs(title = sprintf("Gen %s: No data", gen))
    }
  })
  
  do.call(grid.arrange, c(plots, ncol = 3))
  ggsave(output_file, width = 22, height = 18)
}



# Example usage
database_path <- '/Users/shashankpritam/github/Insertion-Bias-TE/Simulation-Results_Files/simulation_storm/storm/reStorm1.duckdb'
output_file_path <- '/Users/shashankpritam/github/Insertion-Bias-TE/Simulation-Results_Files/simulation_storm/storm/debug_reStorm1.pdf'
plot_data(database_path, output_file_path)

File: Simulation-Results_Files/simulation_storm/storm/debug_plot.py
import re
import pandas as pd
import duckdb
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from matplotlib import gridspec

# Set up the regular expression pattern for data parsing
pattern = re.compile(r'([0-9.-]+)\((-?[0-9]+)\)')

plt.rcParams.update({
    "text.usetex": True,
    "font.family": "serif",
    "font.serif": ["Computer Modern Roman"],
    "font.size": 14
})

def parse_3tot(df, y_col_series, z_col_series):
    def extract_values(series):
        y_vals = pd.Series([np.nan] * len(series))
        z_vals = pd.Series([np.nan] * len(series))
        for i, cell in enumerate(series):
            if isinstance(cell, str):
                parts = cell.split(',')
                for part in parts:
                    match = pattern.match(part)
                    if match:
                        value, key = float(match.group(1)), int(match.group(2))
                        if key == y_col_series[i]:
                            y_vals[i] = value
                        elif key == z_col_series[i]:
                            z_vals[i] = value
        return y_vals, z_vals

    y_vals, z_vals = extract_values(df['3tot'])
    return df.assign(tot_y=y_vals, tot_z=z_vals)

def fetch_and_process_data(database, gen):
    print(f"Fetching data for generation {gen} from {database}")
    conn = duckdb.connect(database=database, read_only=True)
    query = f"""
        SELECT y, z, "3tot", gen
        FROM simulations
        WHERE gen = {gen} AND y BETWEEN -50 AND -43 AND z BETWEEN 47 AND 50

        ORDER BY gen;
    """
    df = conn.execute(query).fetch_df()
    if not df.empty:
        parsed_data = parse_3tot(df, df['y'], df['z'])
        # Directly using the parsed data without mean aggregation
        print("Data for current generation:")
        print(parsed_data[['y', 'z', 'tot_y', 'tot_z']])
        return parsed_data
    return pd.DataFrame()

def scale_data(pivot_y, pivot_z):
    s_values = np.where(pivot_z + pivot_y != 0, 2 * (pivot_y / (pivot_z + pivot_y)) - 1, 0)
    print("Scaled values:\n", s_values)
    return pd.DataFrame(s_values, index=pivot_y.index, columns=pivot_y.columns)

def plot_data(database, output_file):
    plt.figure(figsize=(22, 18))
    gs = gridspec.GridSpec(4, 3)
    generations = [1, 500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000]

    for i, gen in enumerate(generations):
        ax = plt.subplot(gs[i // 3, i % 3])
        gen_data = fetch_and_process_data(database, gen)

        if not gen_data.empty:
            print(f"Gen {gen} - tot_y and tot_z values:")
            print(gen_data[['y', 'z', 'tot_y', 'tot_z']])
            pivot_y = gen_data.pivot(index='y', columns='z', values='tot_y').fillna(0)
            pivot_z = gen_data.pivot(index='y', columns='z', values='tot_z').fillna(0)
            scaled_data = scale_data(pivot_y, pivot_z)
            sns.heatmap(scaled_data, cmap="RdBu_r", center=0, ax=ax, vmin=-1, vmax=1)

        ax.set_title(f'Gen {gen}', fontsize=20)
        ax.set_xlabel('Bias 1 TE', fontsize=20)
        ax.set_ylabel('Bias 2 TE' if (i % 3 == 0) else "", fontsize=20)
        ax.invert_yaxis()
        ax.grid(True, color='lightgrey', linestyle='dotted', linewidth=0.15)

    plt.tight_layout()
    plt.savefig(output_file)
    print(f"Saved plot to {output_file}")

if __name__ == '__main__':
    database = 'reStorm1.duckdb'
    output_file = 'debug_reStorm1.pdf'
    plot_data(database, output_file)
